### PR TITLE
[3135] Add new overflow strategies for node labels

### LINKED ISSUE #3135: Add the ability to specified an overflow strategy on node label

### PR DESCRIPTION
**Pull request by frouene**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3135

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/compatibility/backend/sirius-components-compatibility-emf/src/test/java/org/eclipse/sirius/components/compatibility/emf/compatibility/operations/CreateViewOperationHandlerTests.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
 package org.eclipse.sirius.components.compatibility.emf.compatibility.operations;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
```
Comment (gcoutable) [inline packages/compatibility/backend/sirius-components-compatibility-emf/src/test/java/org/eclipse/sirius/components/compatibility/emf/compatibility/operations/CreateViewOperationHandlerTests.java]: If you want to replace assert statements, replace them with `org.assertj.core.api.Assertions.assertThat`

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/InsideLabel.java [hunk 2]
```diff
@@ -138,6 +146,11 @@ public Builder displayHeaderSeparator(boolean displayHeaderSeparator) {
             return this;
         }
 
+        public Builder overflowStrategy(LabelOverflowStrategy overflowStrategy) {
+            this.overflowStrategy = overflowStrategy;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/InsideLabel.java]: Objects.requireNonNull

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/OutsideLabel.java [hunk 3]
```diff
@@ -17,6 +17,6 @@
  *
  * @author gcoutable
  */
-public record OutsideLabel(String id, String text, OutsideLabelLocation outsideLabelLocation, LabelStyle style) {
+public record OutsideLabel(String id, String text, OutsideLabelLocation outsideLabelLocation, LabelStyle style, LabelOverflowStrategy overflowStrategy) {
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/OutsideLabel.java:22]: RequireNonNull, as well as other attributes ?

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/EdgeComponent.java [hunk 4]
```diff
@@ -303,6 +303,11 @@ private ViewModifier computeState(List<IDiagramEvent> diagramEvents, Element sou
         return state;
     }
 
+    private boolean isElementHidden(HideDiagramElementEvent diagramEvent, String elementId, ViewModifier state) {
+        return (diagramEvent.getElementIds().contains(elementId) && diagramEvent.hideElement())
+                || (!diagramEvent.getElementIds().contains(elementId) && state == ViewModifier.Hidden);
+    }
+
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/EdgeComponent.java]: Dead code

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/InsideLabelDescription.java [hunk 5]
```diff
@@ -149,6 +160,11 @@ public Builder insideLabelLocation(InsideLabelLocation insideLabelLocation) {
             return this;
         }
 
+        public Builder overflowStrategy(LabelOverflowStrategy overflowStrategy) {
+            this.overflowStrategy = overflowStrategy;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/InsideLabelDescription.java]: Objects.requireNonNull

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/OutsideLabelDescription.java [hunk 6]
```diff
@@ -125,13 +134,19 @@ public Builder outsideLabelLocation(OutsideLabelLocation outsideLabelLocation) {
             return this;
         }
 
+        public Builder overflowStrategy(LabelOverflowStrategy overflowStrategy) {
+            this.overflowStrategy = overflowStrategy;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/OutsideLabelDescription.java]: Objects.requireNonNull

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/elements/InsideLabelElementProps.java [hunk 7]
```diff
@@ -129,6 +138,11 @@ public Builder displayHeaderSeparator(boolean displayHeaderSeparator) {
             return this;
         }
 
+        public Builder overflowStrategy(LabelOverflowStrategy overflowStrategy) {
+            this.overflowStrategy = overflowStrategy;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/elements/InsideLabelElementProps.java]: Objects.requireNonNull

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/elements/OutsideLabelElementProps.java [hunk 8]
```diff
@@ -103,12 +112,18 @@ public Builder style(LabelStyle style) {
             return this;
         }
 
+        public Builder overflowStrategy(LabelOverflowStrategy overflowStrategy) {
+            this.overflowStrategy = overflowStrategy;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/elements/OutsideLabelElementProps.java]: Objects.requireNonNull

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/renderer/DiagramElementFactory.java [hunk 9]
```diff
@@ -45,134 +45,137 @@ public class DiagramElementFactory implements IElementFactory {
...
-        } else if (EdgeElementProps.TYPE.equals(type) && props instanceof EdgeElementProps edgeElementProps) {
-            object = this.instantiateEdge(edgeElementProps, children);
-        } else if (LabelElementProps.TYPE.equals(type) && props instanceof LabelElementProps labelElementProps) {
-            object = this.instantiateLabel(labelElementProps);
-        } else if (InsideLabelElementProps.TYPE.equals(type) && props instanceof InsideLabelElementProps insideLabelElementProps) {
-            object = this.instantiateInsideLabel(insideLabelElementProps);
-        } else if (OutsideLabelElementProps.TYPE.equals(type) && props instanceof OutsideLabelElementProps outsideLabelElementProps) {
-            object = this.instantiateOutsideLabel(outsideLabelElementProps);
-        }
-        return object;
```
Comment (gcoutable) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/renderer/DiagramElementFactory.java:62]: Are we sure we want that change?

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/FreeFormNodeLayoutHandler.ts [hunk 10]
```diff
@@ -147,7 +132,7 @@ export class FreeFormNodeLayoutHandler implements INodeLayoutHandler<FreeFormNod
     const northBorderNodeFootprintWidth = getNorthBorderNodeFootprintWidth(visibleNodes, borderNodes, previousDiagram);
     const southBorderNodeFootprintWidth = getSouthBorderNodeFootprintWidth(visibleNodes, borderNodes, previousDiagram);
     const labelOnlyWidth =
-      rectangularNodePadding + (labelElement?.getBoundingClientRect().width ?? 0) + rectangularNodePadding;
+      rectangularNodePadding * 2 + getInsideLabelWidthConstraint(node.data.insideLabel, labelElement);
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/FreeFormNodeLayoutHandler.ts]: Can you reverse the `* 2`?
I know it is twice the `regularNodePadding` but when I wrote that code, it has been written to explain that the `labelWidth` is a `rectangularNodePadding` then, the _label width_, and then, another `rectangularNodePadding`.
With `* 2` you kinda lose that.

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx [hunk 11]
```diff
@@ -94,6 +88,7 @@ export const prepareLayoutArea = (
         id: `${node.id}-label-${index}`,
         key: `${node.id}-label-${index}`,
         role: 'button', // role applied by react flow
+        style: node.data.insideLabel?.overflowStrategy === 'NONE' ? undefined : { maxWidth: node.width },
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx:91]: Why did you add this line?

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx [hunk 12]
```diff
@@ -113,6 +108,7 @@ export const prepareLayoutArea = (
         id: `${outsideLabel.id}-label`,
         key: `${outsideLabel.id}-label`,
         role: 'button', // role applied by react flow
+        style: { maxWidth: node.width },
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx:111]: Why did you add this line?

**Review by gcoutable — DISMISSED** [review]

File: packages/compatibility/backend/sirius-components-compatibility-emf/src/test/java/org/eclipse/sirius/components/compatibility/emf/compatibility/operations/CreateViewOperationHandlerTests.java (hunk 1 above)
Comment (frouene) [reply packages/compatibility/backend/sirius-components-compatibility-emf/src/test/java/org/eclipse/sirius/components/compatibility/emf/compatibility/operations/CreateViewOperationHandlerTests.java]: Done

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/renderer/DiagramElementFactory.java (hunk 9 above)
Comment (frouene) [reply packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/renderer/DiagramElementFactory.java:62]: It's to reduce the cyclomatic complexity.
This pattern with return switch, it's something we already use in other places

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx (hunk 11 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx:91]: To compute the vertical space use if label is wrap, I need to fix the parent div width

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx (hunk 12 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx:111]: same reason

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx (hunk 11 above)
Comment (gcoutable) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layout.tsx:91]: The pre-layout should as much as possible run in the same context as Reactflow.
Here, you are simulating labels inside a node.
I think we should not make that kind of assumption here.

The label should be wrapped on the pre-layout of node.
That means we should never retrieve label bounds from label pre-layout but from node pre-layout.

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeChildrenProvider.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            String nodeId, ViewModifier state, ViewModifier parentState, INodeDescriptionRequestor nodeDescriptionRequestor) {
+        List<Element> nodeChildren = new ArrayList<>();
+
+        nodeChildren.addAll(this.getInsideLabel(nodeVariableManager, nodeDescription, nodeId));
+        nodeChildren.addAll(this.getOutsideLabel(nodeVariableManager, nodeDescription, nodeId));
+        nodeChildren.addAll(this.getBorderNodes(optionalPreviousNode, nodeVariableManager, nodeId, state, nodeDescriptionRequestor));
+        nodeChildren.addAll(this.getChildNodes(optionalPreviousNode, nodeVariableManager, nodeId, parentState, nodeDescriptionRequestor));
+
+        return nodeChildren;
+    }
```
Comment (sbegaudeau) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeChildrenProvider.java]: Technically, this is exactly equivalent to this:

```
const NodeComponent = (props) => {
    const children = nodeChildrenProvider(props);
    return (
        <node>
           {children}
        </node>
    );
}
```

You have a component and instead of having child components, you are using a function which takes props and return elements to compute the children. Could you please provide an API like this instead:

```
const NodeComponent = (props) => {
    return (
       <node>
           <NodeChildrenProvider {...props} />
       </node>
    );
}
```

It would be odd to have a function in React taking props and returning elements but not use it as a component, it is odd here too. And NodeChildrenProvider can still use NodeComponentProps, no issue with that. We would probably not do that in the frontend but it is easier in TypeScript to destructure props and create new props. Given the role of NodeChildrenProvider, handling a part of the concern of NodeComponent, it does make sense to just keep the same props

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/useArrangeAll.ts [hunk 14]
```diff
@@ -50,6 +50,16 @@ const elkOptions = {
   'elk.layered.spacing.edgeNodeBetweenLayers': '40',
 };
 
+const computeHeaderVerticalFootprint = (node, viewportZoom: number): number => {
+  if (node && node.data.insideLabel?.isHeader) {
+    const label = document.querySelector<HTMLDivElement>(`[data-id="${node.data.insideLabel.id}-content"]`);
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/useArrangeAll.ts]: Wouldn't that create errors with diagram in a portal? Something like `reactFlowDomElement.querySelector(...)` would be better to not start from the root of the web page. I think we had this issue somewhere else in the code of Reactflow (in the palette IIRC).

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I have not tested the code but I have one or two small comment on the PR. If it's ok for @gcoutable and if @frouene can rebase it, we can integrate it I think.

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/useArrangeAll.ts (hunk 14 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/useArrangeAll.ts]: I changed it to use the ref of the `DiagramRender`

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeChildrenProvider.java (hunk 13 above)
Comment (frouene) [reply packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeChildrenProvider.java]: I changed from `NodeChildrenProvider` to a fragment `NodeChildrenComponent`

**Comment by sbegaudeau** [comment]
I've just rebased the PR and squashed both commits since the code looks good but it doesn't seems to compile anymore on master...

**Comment by frouene** [comment]
The compilation error has been fixed

**Review by sbegaudeau — APPROVED** [review]
